changefeedccl: implement json schema for enriched envelopes

asg0451 opened the pull request:
Implement a kafka-connect-compatible JSON schema
for the enriched changefeed envelope.

Fixes: #139658

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142280)
<!-- Reviewable:end -->

asg0451 committed:
changefeedccl: implement json schema for enriched envelopes

Implement a kafka-connect-compatible JSON schema
for the enriched changefeed envelope.

Fixes: #139658

Release note: None

asg0451 commented on pkg/ccl/changefeedccl/kcjsonschema/kafka_connect_json_schema.go:
```diff
@@ -0,0 +1,225 @@
+	// TODO: not sure what else to do here. we can make it more dynamic .. but i
+	// don't really think we need to. for reference, other systems will just
+	// stringify json fields, which frankly makes more sense.
+	SchemaTypeOpaqueJSON schemaType = "json"
```
cc @rohan-joshi

asg0451 commented on pkg/ccl/changefeedccl/kcjsonschema/kafka_connect_json_schema.go:
```diff
@@ -0,0 +1,225 @@
+	schemaNameDate      schemaName = "date"
+	schemaNameTime      schemaName = "time"
+	schemaNameTimestamp schemaName = "timestamp"
+	// TODO: These two are additions to the spec that we don't need to make if we don't want to.
```
cc @rohan-joshi

aerfrei commented on pkg/ccl/changefeedccl/enriched_source_provider.go:
```diff
@@ -71,15 +64,60 @@ func (p *enrichedSourceProvider) GetJSON(row cdcevent.Row) (json.JSON, error) {
+		return nil, err
+	}
+
+	return j, nil
```
Was there a difference between writing this out and returning b.Build() directly? Are there times where both j and the error were non-nil?

asg0451 replied:
nope and i couldnt tell you why i did this

aerfrei commented on pkg/ccl/changefeedccl/encoder_json.go:
```diff
@@ -524,12 +575,53 @@ func inSet[S ~string](k S, set map[S]struct{}) bool {
+		source, _ = ptr(e.enrichedEnvelopeSourceProvider.KafkaConnectJSONSchema(), nil)
+	}
+
+	envelope := kcjsonschema.NewEnrichedEnvelope(before, after, source).AsJSON()
```
Are the words envelope and schema synonyms in this context? Should this variable and the NewEnrichedEnvelope reference schemas instead of envelopes?

asg0451 replied:
it's in a package named schema and returns something of type schema. see the guidance in https://go.dev/doc/effective_go#package-names

likewise appending `Schema` to before, after, and source i dont think adds much clarity, esp since we're in makeValueSchema etc

aerfrei commented on pkg/ccl/changefeedccl/enriched_source_provider.go:
```diff
@@ -40,26 +41,18 @@ func newEnrichedSourceProvider(
-		return nil, err
-	}
-	return rec, nil
+func (p *enrichedSourceProvider) KafkaConnectJSONSchema() kcjsonschema.Schema {
```
What is kafka connect? It seems like this is the only kind of JSON schema we emit and we would emit it even for non-kafka sinks if schema is specified. Is that a sensible use case?

aerfrei reviewed (approved):
Looks good! Trying to wrap my head around it and left a few comments.

asg0451 commented on pkg/ccl/changefeedccl/encoder.go:
```diff
@@ -64,3 +66,23 @@ func getEncoder(
 }
+
+// Get the raw SQL-formatted string for a table name
+func getTableName(
```
moved from avro so we can reuse it in json

andyyang890 commented on pkg/ccl/changefeedccl/enriched_source_provider.go:
```diff
@@ -163,9 +150,115 @@ func (p *enrichedSourceProvider) GetJSON(updated cdcevent.Row) (json.JSON, error
+var avroFields []*avro.SchemaField
+
+// filled in by init() using allFieldInfo
+var kcjSchema kcjsonschema.Schema
```
nit: would prefer to see a longer and more descriptive name like `kafkaConnectJSONSchema`

andyyang890 commented on pkg/ccl/changefeedccl/enriched_source_provider.go:
```diff
@@ -163,9 +150,115 @@ func (p *enrichedSourceProvider) GetJSON(updated cdcevent.Row) (json.JSON, error
+}
+
+var allFieldInfo = map[string]fieldInfo{
+	"job_id": {
```
nit: maybe these field names should be constants, mostly to provide peace of mind that we're using the same name in the avro and kafka connect versions of the schema?

andyyang890 commented on pkg/ccl/changefeedccl/encoder_json.go:
```diff
@@ -170,7 +184,11 @@ func makeJSONEncoder(
+	targets                                                                changefeedbase.Targets
 	valueBuilder                                                           *json.FixedKeysObjectBuilder
+	// Cache of the schemas of keys, for use in enriched envelopes with schemas.
+	// type: tableIDAndVersion{} -> json.JSON
```
nit: remove the `{}`

andyyang890 commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -3915,22 +4039,23 @@ func TestChangefeedEnriched(t *testing.T) {
-				if slices.Contains(tc.enrichedProperties, "schema") {
-					// TODO(#139658): add the schema to the key and the value here
-					msg = fmt.Sprintf(`%s: {"payload": {"a": 0}}->{"payload": {"after": {"a": 0, "b": "dog"}, "op": "c"}}`, topic)
+				var jobID int64
```
`jobID` appears to be unused - is that intentional?

andyyang890 commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -3882,18 +3885,139 @@ func TestChangefeedBareAvro(t *testing.T) {
+		{
+			name: "with nothing",
+			messageWithoutSource: map[string]any{
+				// TODO: this is always the same now, so we can simplify these cases.
```
Is there anything that still needs to be done to address this TODO?

asg0451 replied:
thx, missed this during rebase conflict resolution

andyyang890 commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -3882,18 +3885,139 @@ func TestChangefeedBareAvro(t *testing.T) {
+			messageWithoutSource: map[string]any{
+				// TODO: this is always the same now, so we can simplify these cases.
+				"after": afterVal,
+				"op":    "c",
```
Wanted to check: `op` is always included if the user specifies `envelope=enriched`?

asg0451 commented on pkg/ccl/changefeedccl/kcjsonschema/kafka_connect_json_schema.go:
```diff
@@ -0,0 +1,225 @@
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package kcjsonschema
```
hmm that would be more aesthetically appealing, let me think about it. my only hesitation is i dont have any notion that we're going to add anything else to do with kafka connect besides this

asg0451 commented on pkg/ccl/changefeedccl/kcjsonschema/kafka_connect_json_schema.go:
```diff
@@ -0,0 +1,225 @@
+	schemaNameGeography schemaName = "geography"
+)
+
+// Schema is the JSON representation of a Kafka Connect JSON Schema.
```
Unfortunately such a thing does not appear to exist.

andyyang890 commented on pkg/ccl/changefeedccl/kcjsonschema/kafka_connect_json_schema_test.go:
```diff
@@ -0,0 +1,384 @@
+
+		serializedDatum := tree.Serialize(datum)
+		// name can be "char" (with quotes), so needs to be escaped.
+		escapedName := fmt.Sprintf("%s_table", strings.Replace(typ.String(), "\"", "", -1))
```
nit: given the updated comment, maybe this should be named quoteStrippedName or something

andyyang890 reviewed (approved):
LGTM

asg0451 commented:
ty @andyyang890

asg0451 commented:
bors r=andyyang890
